File: 297. Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        result = []
        queue = [root]
        while any(queue):
            result += [str(node.val) if node else 'null' for node in queue]
            temp = []
            for node in queue:
                if node:
                    temp.append(node.left)
                    temp.append(node.right)
            queue = temp
        s = '[' + ','.join(result) + ']'
        return s
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        nodelist = data[1:-1].split(',')
        if not nodelist[0]:
            return None
        while nodelist[-1] == 'null':
            nodelist.pop()
        count = 1
        root = TreeNode(int(nodelist[0]))
        ls = [root]
        while count<len(nodelist):
            temp = []
            for node in ls:
                if node:
                    node.left = TreeNode(int(nodelist[count])) if nodelist[count]!='null' else None
                    count += 1
                    if count >= len(nodelist):
                        break
                    temp.append(node.left)
                    node.right = TreeNode(int(nodelist[count])) if nodelist[count]!='null' else None
                    count += 1
                    if count >= len(nodelist):
                        break
                    temp.append(node.right)
            ls = temp
        return root

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: 297. Serialize and Deserialize Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    public String serialize(TreeNode root) {
        String n = "null", comma = ",";
        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> dq = new LinkedList<>();
        dq.offer(root);
        while (!dq.isEmpty()) {
            int size = dq.size();
            for (int i=0;i<size;i++) {
                TreeNode node = dq.poll();
                if (node == null) {
                    sb.append(n);
                    sb.append(comma);
                } else {
                    sb.append(Integer.toString(node.val));
                    sb.append(comma);
                    dq.offer(node.left);
                    dq.offer(node.right);
                }
            }
        }
        return sb.toString();
    }

    public TreeNode deserialize(String data) {
        String[] strs = data.split(",");
        Queue<TreeNode> dq = new LinkedList<>();
        if (strs[0].equals("null")) {
            return null;
        }
        TreeNode root = new TreeNode(Integer.parseInt(strs[0]));
        dq.offer(root);
        int pos = 1;
        while (!dq.isEmpty() && pos < strs.length) {
            TreeNode node = dq.poll();
            String left = strs[pos++];
            if (!left.equals("null")) {
                TreeNode nl = new TreeNode(Integer.parseInt(left));
                node.left = nl;
                dq.offer(nl);
            }
            String right = strs[pos++];
            if (!right.equals("null")) {
                TreeNode nr = new TreeNode(Integer.parseInt(right));
                node.right = nr;
                dq.offer(nr);
            }
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));